### PR TITLE
Feature/update to angular 16

### PR DESCRIPTION
**Pull request by Iloveall**
## Description

Fixes #[1368](https://github.com/France-ioi/AlgoreaFrontend/issues/1368)

## Notes (out of scope, known isues, hints for reviewing code, ...)  (optional)

Branch for [test](https://dev.algorea.org/branch/feature/update-to-angular-16/en/a/home;pa=0)

I've realised that after design the scroll to nav item feature has been brokne. So I've fixed it (seems need to improve it after all our changes)

For some reason the primeng has removed the next generic type [https://github.com/primefaces/primeng/commit/d3b00fcd6f7cfe30d1261fa893211062d9fa26f6#diff-d92d38111264028992d6cb71[…]9510846b1d3c44187c32ff979f3R13](https://github.com/primefaces/primeng/commit/d3b00fcd6f7cfe30d1261fa893211062d9fa26f6#diff-d92d38111264028992d6cb71dde1070be12aa9510846b1d3c44187c32ff979f3R13)

Fixed with custom interface

Replaced deprecated scrollbar lib https://www.npmjs.com/package/ngx-perfect-scrollbar with https://www.npmjs.com/package/ngx-scrollbar

## Test cases

- [ ] Case 1:
  1. Given I am the usual user
  2. When I go to [this page](https://dev.algorea.org/branch/feature/update-to-angular-16/en/a/7132714098371307259;p=;a=0)
  3. And I click on prev button of the neighbours widget
  5. Then I see scroll to with animation in left menu

### CODE REVIEW

File: src/app/core/app.module.ts [hunk 1]
```diff
@@ -75,13 +74,13 @@ const DEFAULT_PERFECT_SCROLLBAR_CONFIG: PerfectScrollbarConfigInterface = {
   ],
   imports: [
     BrowserModule,
-    LetModule,
+    LetDirective,
```
Comment (Iloveall) [inline src/app/core/app.module.ts:77]: Deprecated

File: src/app/shared/guards/before-unload-guard.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ export interface BeforeUnloadComponent {
 }
 
 @Injectable()
-export class BeforeUnloadGuard implements CanDeactivate<BeforeUnloadComponent> {
```
Comment (Iloveall) [inline src/app/shared/guards/before-unload-guard.ts:13]: Has been removed by cli as it deprecated

File: src/app/shared/guards/pending-changes-guard.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ export interface PendingChangesComponent {
 @Injectable({
   providedIn: 'root'
 })
-export class PendingChangesGuard implements CanDeactivate<PendingChangesComponent>, OnDestroy {
+export class PendingChangesGuard implements OnDestroy {
```
Comment (smadbe) [inline src/app/shared/guards/pending-changes-guard.ts:17]: same here

Comment (Iloveall) [reply src/app/shared/guards/pending-changes-guard.ts:17]: Tested

File: src/app/shared/services/layout.service.ts [hunk 4]
```diff
@@ -114,7 +114,7 @@ export class LayoutService implements OnDestroy {
 @Injectable({
   providedIn: 'root'
 })
-export class DefaultLayoutInitService implements CanActivate {
+export class DefaultLayoutInitService {
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/shared/services/layout.service.ts:117]: Tested

File: src/app/shared/guards/before-unload-guard.ts (hunk 2 above)
Comment (Iloveall) [reply src/app/shared/guards/before-unload-guard.ts:13]: Tested

**Pull request merged**